### PR TITLE
ITE drivers/kscan: add keyboard scan driver for it8xxx2_evb

### PR DESCRIPTION
**Pull request by RuibinChang**
Add keyboard scan driver for board it8xxx2_evb.

Signed-off-by: Ruibin Chang <ruibin.chang@ite.com.tw>

---
This change is [Reviewable](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/38169)

### CODE REVIEW

File: drivers/kscan/Kconfig.it8xxx2 [hunk 1]
```diff
@@ -0,0 +1,45 @@
...
+	depends on SOC_IT8XXX2
+	select MULTITHREADING
+	help
+	  This option enables the ITE keyboard scan driver.
+
+if KSCAN_ITE_IT8XXX2
+
+config KSCAN_ITE_IT8XXX2_COLUMN_SIZE
+	int "KSCAN_ITE_IT8XXX2_COLUMN_SIZE"
+	default 16
```
[the remark this answers is not in the document]
Comment (RuibinChang) [reply drivers/kscan/Kconfig.it8xxx2:16]: Done

File: drivers/kscan/Kconfig.it8xxx2 [hunk 2]
```diff
@@ -0,0 +1,45 @@
+# Copyright (c) 2021 ITE Corporation. All Rights Reserved.
+# SPDX-License-Identifier: Apache-2.0
+
+menuconfig KSCAN_ITE_IT8XXX2
+	bool "ITE KSCAN driver"
+	depends on SOC_IT8XXX2
```
[the remark this answers is not in the document]
Comment (RuibinChang) [reply drivers/kscan/Kconfig.it8xxx2]: Done

File: drivers/kscan/kscan_ite_it8xxx2.c [hunk 3]
```diff
@@ -0,0 +1,572 @@
...
+	volatile uint8_t *reg_wuesr3 = (uint8_t *)config->reg_wuesr3;
+	volatile uint8_t *reg_wuenr3 = (uint8_t *)config->reg_wuenr3;
+	struct kscan_it8xxx2_data *data = DRV_DATA(dev);
+	struct kbs_reg *const kbs_reg = DRV_REG(dev);
+
+	/* Disable wakeup and interrupt of KSI pins before configuring */
+	keyboard_raw_enable_interrupt(dev, 0);
+
+	/* Bit[2] = 1: Enable the internal pull-up of KSO[15:0] pins */
+	kbs_reg->KBS_KSOCTRL = IT8XXX2_KBS_KSOPU;
```
[the remark this answers is not in the document]
Comment (RuibinChang) [reply drivers/kscan/kscan_ite_it8xxx2.c]: Bit2: PU of KSO[0:15], it only enable 16 pins. 
PU of KSO16 and KSO17 are controlled by GPCRC3 and GPCRC5 registers.

By the way, I found that it didn't set open drain for KSO[15:0] in below setting when column size is 16.
	/* Bit[2] = 1: Enable the internal pull-up of KSO[15:0] pins */
	kbs_reg->KBS_KSOCTRL = IT8XXX2_KBS_KSOPU;
#if (CONFIG_KSCAN_ITE_IT8XXX2_COLUMN_SIZE > 16)
	/* Bit[0] = 1: Enable the open-drain mode of KSO[17:0] pins */
	kbs_reg->KBS_KSOCTRL |= IT8XXX2_KBS_KSOOD;
...
#endif
So I change to kbs_reg->KBS_KSOCTRL = (IT8XXX2_KBS_KSOOD | IT8XXX2_KBS_KSOPU); 
and this is okay for KSO16 and KSO17 when column size is 16.

**Comment by RuibinChang** [comment]
Any other comments?

**Review by brockus-zephyr — APPROVED** [review]
> Reviewed 8 of 10 files at r1, 2 of 2 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/38169)* status: all files reviewed, 3 unresolved discussions (waiting on @albertofloyd, @carlescufi, @galak, @GTLin08, @jackrosenthal, @katsuster, @keith-zephyr, @kgugala, @MaureenHelm, @mbolivar-nordic, @mgielda, @nashif, @pgielda, @sjg20, and @VenkatKotakonda)

**Review by keith-zephyr — CHANGES_REQUESTED** [review]
> Reviewed 4 of 10 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/38169)* status: all files reviewed, 14 unresolved discussions (waiting on @albertofloyd, @carlescufi, @galak, @GTLin08, @jackrosenthal, @katsuster, @kgugala, @MaureenHelm, @mbolivar-nordic, @mgielda, @nashif, @pgielda, @RuibinChang, @sjg20, and @VenkatKotakonda)

___
*[boards/riscv/it8xxx2_evb/it8xxx2_evb.dts, line 43 at r1](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/38169#-Mic_7f75ZdPx8XprvB0:-Mic_7f75ZdPx8XprvB1:b7ercek) ([raw file](https://github.com/zephyrproject-rtos/zephyr/blob/128a1d94d96af50d2fa94e61c6196076ba523a8d/boards/riscv/it8xxx2_evb/it8xxx2_evb.dts#L43)):*
<details><summary><i alt="summary">Quoted 16 lines of code…</i></summary>

> ```
> 	soc {
> 		kscan0: kscan@f01d00 {
> 			compatible = "ite,it8xxx2-kscan";
> 			reg = <0x00f01d00 0x29
> 			       0x00f01b02 0x01   /* WUEMR3 */
> 			       0x00f01b06 0x01   /* WUESR3 */
> 			       0x00f01b0a 0x01>; /* WUENR3 */
> 			label = "KSCAN";
> 			interrupt-parent = <&intc>;
> 			interrupts = <IT8XXX2_IRQ_WKINTC IRQ_TYPE_LEVEL_HIGH>;
> 			status = "disabled";
> 			gpio-dev = <&gpioc>;
> 			pinctrl-0 = <&pinctrl_kso16   /* GPC3 */
> 				     &pinctrl_kso17>; /* GPC5 */
> 		};
> 	};
> ```
</details>

I think node belongs in it8xxx2.dtsi instead of in the eval board.
Also, the wake up controller probably needs it's own driver, with the keyboard scan node specifying the wake up group.  But this can be added as a follow on PR.
___
*[drivers/kscan/kscan_ite_it8xxx2.c, line 74 at r2](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/38169#-MigJpqz7_vWnU52_fTO:-MigJpr-Bv532KYgAjIL:biex5gb) ([raw file](https://github.com/zephyrproject-rtos/zephyr/blob/9fab19fcc97a43117cb985423a013603edea1f83/drivers/kscan/kscan_ite_it8xxx2.c#L74)):*
> ```C
> 	uint32_t deb_time_press;
> 	uint32_t deb_time_rel;
> 	int64_t poll_timeout;
> ```

What's the reason for making poll_timeout a int64_t instead of int32_t?  The poll_expired() routine looks odd with the mix of 32-bit and 64-bit values.
___
*[drivers/kscan/kscan_ite_it8xxx2.c, line 147 at r2](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/38169#-MigCTl1CK6kWCxeFCVy:-MigCTl1CK6kWCxeFCVz:bevxxzn) ([raw file](https://github.com/zephyrproject-rtos/zephyr/blob/9fab19fcc97a43117cb985423a013603edea1f83/drivers/kscan/kscan_ite_it8xxx2.c#L147)):*
> ```
> MAX_MATRIX_KEY_COLS
> ```
CONFIG_KSCAN_ITE_IT8XXX2_COLUMN_SIZE?
___
*[drivers/kscan/kscan_ite_it8xxx2.c, line 157 at r2](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/38169#-MigBP8b7fH5eUFA22r7:-MigBP8c85QMS4XrExnN:b6chen5) ([raw file](https://github.com/zephyrproject-rtos/zephyr/blob/9fab19fcc97a43117cb985423a013603edea1f83/drivers/kscan/kscan_ite_it8xxx2.c#L157)):*
<details><summary><i alt="summary">Quoted 8 lines of code…</i></summary>

> ```
> 			/* we and the columns to detect a "block".
> 			 * this is an indication of ghosting, due to current
> 			 * flowing from a key which was never pressed. in our
> 			 * case, current flowing is a bit set to 1 as we
> 			 * flipped the bits when the matrix was scanned.
> 			 * now we or the colums using z&(z-1) which is
> 			 * non-zero only if z has more than one bit set.
> 			 */
> ```
</details>

Suggest capitalizing AND, OR, in the above comment to make it clear you are talking about logical operations.
For example: "We AND the columns to detect a...",  "Now we OR the columns".
___
*[drivers/kscan/kscan_ite_it8xxx2.c, line 175 at r2](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/38169#-MigBwglEYxvZbX_ZqH2:-MigBwglEYxvZbX_ZqH3:b-k4815) ([raw file](https://github.com/zephyrproject-rtos/zephyr/blob/9fab19fcc97a43117cb985423a013603edea1f83/drivers/kscan/kscan_ite_it8xxx2.c#L175)):*
> ```C
> 	uint8_t key_event = 0U;
> 
> 	for (int col = 0; col < MAX_MATRIX_KEY_COLS; col++) {
> ```

Should this loop be from 0 to CONFIG_KSCAN_ITE_IT8XXX2_COLUMN_SIZE?  drive_keyboard_column() is a no-op if the column is greater than CONFIG_KSCAN_ITE_IT8XXX2_COLUMN_SIZE, but no sense int attempting to scan unused columns.
___
*[drivers/kscan/kscan_ite_it8xxx2.c, line 245 at r2](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/38169#-MigDRRC1F5G-M6yAUGe:-MigDRRC1F5G-M6yAUGf:bju0azt) ([raw file](https://github.com/zephyrproject-rtos/zephyr/blob/9fab19fcc97a43117cb985423a013603edea1f83/drivers/kscan/kscan_ite_it8xxx2.c#L245)):*
> ```
> 	uint8_t row_changed = 0U;
> 	uint8_t deb_col;
> ```
Move variable declaration to the beginning of the function.  I don't think it's common practice in the zephyr code to mix the declarations and code, even though allowed by C99.
___
*[drivers/kscan/kscan_ite_it8xxx2.c, line 252 at r2](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/38169#-MigE-KL1gnTeuLSgwt0:-MigE-KL1gnTeuLSgwt1:bu9mjcc) ([raw file](https://github.com/zephyrproject-rtos/zephyr/blob/9fab19fcc97a43117cb985423a013603edea1f83/drivers/kscan/kscan_ite_it8xxx2.c#L252)):*
> ```
> MAX_MATRIX_KEY_COLS
> ```
CONFIG_KSCAN_ITE_IT8XXX2_COLUMN_SIZE
___
*[drivers/kscan/kscan_ite_it8xxx2.c, line 260 at r2](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/38169#-MigE48uAngOuKcGh9zU:-MigE48uAngOuKcGh9zV:b89vhbe) ([raw file](https://github.com/zephyrproject-rtos/zephyr/blob/9fab19fcc97a43117cb985423a013603edea1f83/drivers/kscan/kscan_ite_it8xxx2.c#L260)):*
> ```
> MAX_MATRIX_KEY_ROWS
> ```
CONFIG_KSCAN_ITE_IT8XXX2_ROW_SIZE
___
*[drivers/kscan/kscan_ite_it8xxx2.c, line 274 at r2](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/38169#-MigEBHz1mW-KPr46_fm:-MigEBHz1mW-KPr46_fn:bu9mjcc) ([raw file](https://github.com/zephyrproject-rtos/zephyr/blob/9fab19fcc97a43117cb985423a013603edea1f83/drivers/kscan/kscan_ite_it8xxx2.c#L274)):*
> ```
> MAX_MATRIX_KEY_COLS
> ```
CONFIG_KSCAN_ITE_IT8XXX2_COLUMN_SIZE
___
*[drivers/kscan/kscan_ite_it8xxx2.c, line 281 at r2](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/38169#-MigEDf1-qKoJM4p3Nc_:-MigEDf1-qKoJM4p3Nca:b89vhbe) ([raw file](https://github.com/zephyrproject-rtos/zephyr/blob/9fab19fcc97a43117cb985423a013603edea1f83/drivers/kscan/kscan_ite_it8xxx2.c#L281)):*
> ```
> MAX_MATRIX_KEY_ROWS
> ```
CONFIG_KSCAN_ITE_IT8XXX2_ROW_SIZE
___
*[drivers/kscan/kscan_ite_it8xxx2.c, line 378 at r2](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/38169#-MigKSD-CTk0Y6GXXEVP:-MigKSD-CTk0Y6GXXEVQ:b-2rjylo) ([raw file](https://github.com/zephyrproject-rtos/zephyr/blob/9fab19fcc97a43117cb985423a013603edea1f83/drivers/kscan/kscan_ite_it8xxx2.c#L378)):*
> ```
> 			} else if (poll_expired(start_poll_cycles,
> 						&local_poll_timeout)) {
> 				break;
> ```
I don't understand the logic here.  If there timeout, the polling task is exited.  What's the use case for this?

**Review by keith-zephyr — CHANGES_REQUESTED** [review]
> Reviewed 4 of 10 files at r1, 1 of 2 files at r2.
> *[Reviewable](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/38169)* status: all files reviewed, 15 unresolved discussions (waiting on @albertofloyd, @carlescufi, @galak, @GTLin08, @jackrosenthal, @katsuster, @kgugala, @MaureenHelm, @mbolivar-nordic, @mgielda, @nashif, @pgielda, @RuibinChang, @sjg20, and @VenkatKotakonda)

___
*[soc/riscv/riscv-ite/common/chip_chipregs.h, line 1002 at r2](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/38169#-MigKv3O70UftbE_FDlt:-MigKv3O70UftbE_FDlu:b-55i898) ([raw file](https://github.com/zephyrproject-rtos/zephyr/blob/9fab19fcc97a43117cb985423a013603edea1f83/soc/riscv/riscv-ite/common/chip_chipregs.h#L1002)):*
> ```objc
>  */
> #ifndef __ASSEMBLER__
> struct kbs_reg {
> ```

I recommend a BUILD_ASSERT that at least checks the size of the keyboard register space.

**Review by RuibinChang — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/38169)* status: 9 of 10 files reviewed, 15 unresolved discussions (waiting on @albertofloyd, @brockus-zephyr, @carlescufi, @galak, @GTLin08, @jackrosenthal, @katsuster, @keith-zephyr, @kgugala, @MaureenHelm, @mbolivar-nordic, @mgielda, @nashif, @pgielda, @RuibinChang, @sjg20, and @VenkatKotakonda)

___
*[boards/riscv/it8xxx2_evb/it8xxx2_evb.dts, line 43 at r1](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/38169#-Mic_7f75ZdPx8XprvB0:-MiueqDr2tOFD6kdlSnm:b2la9yu) ([raw file](https://github.com/zephyrproject-rtos/zephyr/blob/128a1d94d96af50d2fa94e61c6196076ba523a8d/boards/riscv/it8xxx2_evb/it8xxx2_evb.dts#L43)):*
<details><summary><i>Previously, keith-zephyr (Keith Short) wrote…</i></summary><blockquote>

> ```
> 	soc {
> 		kscan0: kscan@f01d00 {
> 			compatible = "ite,it8xxx2-kscan";
> 			reg = <0x00f01d00 0x29
> 			       0x00f01b02 0x01   /* WUEMR3 */
> 			       0x00f01b06 0x01   /* WUESR3 */
> 			       0x00f01b0a 0x01>; /* WUENR3 */
> 			label = "KSCAN";
> 			interrupt-parent = <&intc>;
> 			interrupts = <IT8XXX2_IRQ_WKINTC IRQ_TYPE_LEVEL_HIGH>;
> 			status = "disabled";
> 			gpio-dev = <&gpioc>;
> 			pinctrl-0 = <&pinctrl_kso16   /* GPC3 */
> 				     &pinctrl_kso17>; /* GPC5 */
> 		};
> 	};
> ```
I think node belongs in it8xxx2.dtsi instead of in the eval board.
Also, the wake up controller probably needs it's own driver, with the keyboard scan node specifying the wake up group.  But this can be added as a follow on PR.
</blockquote></details>

Yes, I agree that kscan node belongs chip it8xxx2, but if kscan node move to it8xxx2.dtsi, then building hayato in cros_ec/zephyr shows 
ERROR:/mnt/host/source/src/platform/ec/zephyr/build/packer/source.dts:299.24-308.5: Warning (unique_unit_address): /soc/kscan@f01d00: duplicate unit-address (also used in node /soc/cros-kb-raw@f01d00).   
I think this also need to delete node of hayato and at least change the getting node label string from cros_kb_raw to kscan in shim/keyboard_raw.c. 
How do you think, Keith?

Ok, I will create another PR for WUI.
___
*[drivers/kscan/kscan_ite_it8xxx2.c, line 74 at r2](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/38169#-MigJpqz7_vWnU52_fTO:-MiufEQRDyowMRMYhh24:b-1kw7pf) ([raw file](https://github.com/zephyrproject-rtos/zephyr/blob/9fab19fcc97a43117cb985423a013603edea1f83/drivers/kscan/kscan_ite_it8xxx2.c#L74)):*
<details><summary><i>Previously, keith-zephyr (Keith Short) wrote…</i></summary><blockquote>

What's the reason for making poll_timeout a int64_t instead of int32_t?  The poll_expired() routine looks odd with the mix of 32-bit and 64-bit values.
</blockquote></details>

poll_timeout unit is micro-second, just in case the destination time >2147 sec (when int32_t overflow) , but I think it won't >2147 sec, so I change variable type to int32_t.
___
*[drivers/kscan/kscan_ite_it8xxx2.c, line 147 at r2](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/38169#-MigCTl1CK6kWCxeFCVy:-MiufON0AdKTKodLKeSb:b-8p3dww) ([raw file](https://github.com/zephyrproject-rtos/zephyr/blob/9fab19fcc97a43117cb985423a013603edea1f83/drivers/kscan/kscan_ite_it8xxx2.c#L147)):*
<details><summary><i>Previously, keith-zephyr (Keith Short) wrote…</i></summary><blockquote>

> ```
> MAX_MATRIX_KEY_COLS
> ```
CONFIG_KSCAN_ITE_IT8XXX2_COLUMN_SIZE?
</blockquote></details>

Yes, because the CONFIG_ string is too long, I define another in line 22 and 23 to replace:
#define MAX_MATRIX_KEY_COLS        CONFIG_KSCAN_ITE_IT8XXX2_COLUMN_SIZE
#define MAX_MATRIX_KEY_ROWS       CONFIG_KSCAN_ITE_IT8XXX2_ROW_SIZE
Shall I remove the #define and use CONFIG_*?
___
*[drivers/kscan/kscan_ite_it8xxx2.c, line 157 at r2](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/38169#-MigBP8b7fH5eUFA22r7:-MiufWyp29550Nqs-IIV:bb6un1j) ([raw file](https://github.com/zephyrproject-rtos/zephyr/blob/9fab19fcc97a43117cb985423a013603edea1f83/drivers/kscan/kscan_ite_it8xxx2.c#L157)):*
<details><summary><i>Previously, keith-zephyr (Keith Short) wrote…</i></summary><blockquote>

> ```
> 			/* we and the columns to detect a "block".
> 			 * this is an indication of ghosting, due to current
> 			 * flowing from a key which was never pressed. in our
> 			 * case, current flowing is a bit set to 1 as we
> 			 * flipped the bits when the matrix was scanned.
> 			 * now we or the colums using z&(z-1) which is
> 			 * non-zero only if z has more than one bit set.
> 			 */
> ```
Suggest capitalizing AND, OR, in the above comment to make it clear you are talking about logical operations.
For example: "We AND the columns to detect a...",  "Now we OR the columns".
</blockquote></details>

Done
___
*[drivers/kscan/kscan_ite_it8xxx2.c, line 245 at r2](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/38169#-MigDRRC1F5G-M6yAUGe:-MiufbZ25hMvHvhjQrfk:bb6un1j) ([raw file](https://github.com/zephyrproject-rtos/zephyr/blob/9fab19fcc97a43117cb985423a013603edea1f83/drivers/kscan/kscan_ite_it8xxx2.c#L245)):*
<details><summary><i>Previously, keith-zephyr (Keith Short) wrote…</i></summary><blockquote>

> ```
> 	uint8_t row_changed = 0U;
> 	uint8_t deb_col;
> ```
Move variable declaration to the beginning of the function.  I don't think it's common practice in the zephyr code to mix the declarations and code, even though allowed by C99.
</blockquote></details>

Done
___
*[drivers/kscan/kscan_ite_it8xxx2.c, line 378 at r2](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/38169#-MigKSD-CTk0Y6GXXEVP:-MiufjUGC267w8pAsBfm:b6qv7lv) ([raw file](https://github.com/zephyrproject-rtos/zephyr/blob/9fab19fcc97a43117cb985423a013603edea1f83/drivers/kscan/kscan_ite_it8xxx2.c#L378)):*
<details><summary><i>Previously, keith-zephyr (Keith Short) wrote…</i></summary><blockquote>

> ```
> 			} else if (poll_expired(start_poll_cycles,
> 						&local_poll_timeout)) {
> 				break;
> ```
I don't understand the logic here.  If there timeout, the polling task is exited.  What's the use case for this?
</blockquote></details>

If the key is released and > 100ms, then we go to init KSO state and wait for next pressed event.

**Review by RuibinChang — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/38169)* status: 9 of 10 files reviewed, 15 unresolved discussions (waiting on @albertofloyd, @brockus-zephyr, @carlescufi, @galak, @GTLin08, @jackrosenthal, @katsuster, @keith-zephyr, @kgugala, @MaureenHelm, @mbolivar-nordic, @mgielda, @nashif, @pgielda, @RuibinChang, @sjg20, and @VenkatKotakonda)

___
*[drivers/kscan/kscan_ite_it8xxx2.c, line 175 at r2](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/38169#-MigBwglEYxvZbX_ZqH2:-Miz0_KL3BVcalNOXl_7:bxtcpb1) ([raw file](https://github.com/zephyrproject-rtos/zephyr/blob/9fab19fcc97a43117cb985423a013603edea1f83/drivers/kscan/kscan_ite_it8xxx2.c#L175)):*
<details><summary><i>Previously, keith-zephyr (Keith Short) wrote…</i></summary><blockquote>

Should this loop be from 0 to CONFIG_KSCAN_ITE_IT8XXX2_COLUMN_SIZE?  drive_keyboard_column() is a no-op if the column is greater than CONFIG_KSCAN_ITE_IT8XXX2_COLUMN_SIZE, but no sense int attempting to scan unused columns.
</blockquote></details>

Yes, the loop is indeed from 0 to CONFIG_KSCAN_ITE_IT8XXX2_COLUMN_SIZE.  
There is a definition
#define MAX_MATRIX_KEY_COLS 
     CONFIG_KSCAN_ITE_IT8XXX2_COLUMN_SIZE, so it won't happen that the column is greater than CONFIG_KSCAN_ITE_IT8XXX2_COLUMN_SIZE.
___
*[soc/riscv/riscv-ite/common/chip_chipregs.h, line 1002 at r2](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/38169#-MigKv3O70UftbE_FDlt:-Miz__XuBWLiB5vbnKju:bb6un1j) ([raw file](https://github.com/zephyrproject-rtos/zephyr/blob/9fab19fcc97a43117cb985423a013603edea1f83/soc/riscv/riscv-ite/common/chip_chipregs.h#L1002)):*
<details><summary><i>Previously, keith-zephyr (Keith Short) wrote…</i></summary><blockquote>

I recommend a BUILD_ASSERT that at least checks the size of the keyboard register space.
</blockquote></details>

Done

**Review by brockus-zephyr — APPROVED** [review]
> Reviewed 3 of 3 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/38169)* status: all files reviewed, 15 unresolved discussions (waiting on @albertofloyd, @carlescufi, @galak, @GTLin08, @jackrosenthal, @katsuster, @keith-zephyr, @kgugala, @MaureenHelm, @mbolivar-nordic, @mgielda, @nashif, @pgielda, @RuibinChang, @sjg20, and @VenkatKotakonda)

**Review by keith-zephyr — CHANGES_REQUESTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/38169)* status: all files reviewed, 6 unresolved discussions (waiting on @albertofloyd, @carlescufi, @galak, @GTLin08, @jackrosenthal, @katsuster, @kgugala, @MaureenHelm, @mbolivar-nordic, @mgielda, @nashif, @pgielda, @RuibinChang, @sjg20, and @VenkatKotakonda)

___
*[boards/riscv/it8xxx2_evb/it8xxx2_evb.dts, line 43 at r1](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/38169#-Mic_7f75ZdPx8XprvB0:-MjFAzuXCEAw6PXRwTcJ:b-i6cm4s) ([raw file](https://github.com/zephyrproject-rtos/zephyr/blob/128a1d94d96af50d2fa94e61c6196076ba523a8d/boards/riscv/it8xxx2_evb/it8xxx2_evb.dts#L43)):*
<details><summary><i>Previously, RuibinChang wrote…</i></summary><blockquote>

Yes, I agree that kscan node belongs chip it8xxx2, but if kscan node move to it8xxx2.dtsi, then building hayato in cros_ec/zephyr shows 
ERROR:/mnt/host/source/src/platform/ec/zephyr/build/packer/source.dts:299.24-308.5: Warning (unique_unit_address): /soc/kscan@f01d00: duplicate unit-address (also used in node /soc/cros-kb-raw@f01d00).   
I think this also need to delete node of hayato and at least change the getting node label string from cros_kb_raw to kscan in shim/keyboard_raw.c. 
How do you think, Keith?

Ok, I will create another PR for WUI.
</blockquote></details>

For the Hayato board, you could add "/delete-node/ kscan@f01d00".  This would allow the Hayato board to enable the cros-kb-raw driver.  Because the cros-kb-raw driver is only used for Chrome EC, the upstream Zephyr driver should take priority.
___
*[drivers/kscan/kscan_ite_it8xxx2.c, line 147 at r2](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/38169#-MigCTl1CK6kWCxeFCVy:-MjFCEud8xInamDy_VI_:b-vq0owk) ([raw file](https://github.com/zephyrproject-rtos/zephyr/blob/9fab19fcc97a43117cb985423a013603edea1f83/drivers/kscan/kscan_ite_it8xxx2.c#L147)):*
<details><summary><i>Previously, RuibinChang wrote…</i></summary><blockquote>

Yes, because the CONFIG_ string is too long, I define another in line 22 and 23 to replace:
#define MAX_MATRIX_KEY_COLS        CONFIG_KSCAN_ITE_IT8XXX2_COLUMN_SIZE
#define MAX_MATRIX_KEY_ROWS       CONFIG_KSCAN_ITE_IT8XXX2_ROW_SIZE
Shall I remove the #define and use CONFIG_*?
</blockquote></details>

Yes, I think that would be clearer.  The MAX_MATRIX defines don't really add value.
___
*[drivers/kscan/kscan_ite_it8xxx2.c, line 378 at r2](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/38169#-MigKSD-CTk0Y6GXXEVP:-MjFDeAF7_96YtPz6WFU:bvfrb7y) ([raw file](https://github.com/zephyrproject-rtos/zephyr/blob/9fab19fcc97a43117cb985423a013603edea1f83/drivers/kscan/kscan_ite_it8xxx2.c#L378)):*
<details><summary><i>Previously, RuibinChang wrote…</i></summary><blockquote>

If the key is released and > 100ms, then we go to init KSO state and wait for next pressed event. 
</blockquote></details>

My mistake, I didn't notice that this break only exits the inner while() loop.  The outer while(true) loop will keep the thread active.

**Review by RuibinChang — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/38169)* status: 8 of 12 files reviewed, 6 unresolved discussions (waiting on @albertofloyd, @brockus-zephyr, @carlescufi, @galak, @GTLin08, @jackrosenthal, @katsuster, @keith-zephyr, @kgugala, @MaureenHelm, @mbolivar-nordic, @mgielda, @nashif, @pgielda, @RuibinChang, @sjg20, and @VenkatKotakonda)

___
*[boards/riscv/it8xxx2_evb/it8xxx2_evb.dts, line 43 at r1](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/38169#-Mic_7f75ZdPx8XprvB0:-MjcTA057YwppgOiYUk5:b-896fix) ([raw file](https://github.com/zephyrproject-rtos/zephyr/blob/128a1d94d96af50d2fa94e61c6196076ba523a8d/boards/riscv/it8xxx2_evb/it8xxx2_evb.dts#L43)):*
<details><summary><i>Previously, keith-zephyr (Keith Short) wrote…</i></summary><blockquote>

For the Hayato board, you could add "/delete-node/ kscan@f01d00".  This would allow the Hayato board to enable the cros-kb-raw driver.  Because the cros-kb-raw driver is only used for Chrome EC, the upstream Zephyr driver should take priority.
</blockquote></details>

Done.
___
*[drivers/kscan/kscan_ite_it8xxx2.c, line 147 at r2](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/38169#-MigCTl1CK6kWCxeFCVy:-MjcT4XR7VZlV5YhhhW3:b-896fix) ([raw file](https://github.com/zephyrproject-rtos/zephyr/blob/9fab19fcc97a43117cb985423a013603edea1f83/drivers/kscan/kscan_ite_it8xxx2.c#L147)):*
<details><summary><i>Previously, keith-zephyr (Keith Short) wrote…</i></summary><blockquote>

Yes, I think that would be clearer.  The MAX_MATRIX defines don't really add value.
</blockquote></details>

Done.

**Review by keith-zephyr — APPROVED** [review]
> Reviewed 1 of 3 files at r4, 4 of 4 files at r5, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/zephyrproject-rtos/zephyr/38169)* status: all files reviewed, 4 unresolved discussions (waiting on @albertofloyd, @carlescufi, @galak, @GTLin08, @jackrosenthal, @katsuster, @keith-zephyr, @kgugala, @MaureenHelm, @mbolivar-nordic, @mgielda, @nashif, @pgielda, @sjg20, and @VenkatKotakonda)

**Review by jackrosenthal — APPROVED** [review]

**Review by VenkatKotakonda — APPROVED** [review]

**Comment by RuibinChang** [comment]
Since there are no more comments, could we merge this PR? Thanks!

**Pull request merged**